aten/src/ATen/native/RNN.cpp

```diff
@@ -0,0 +1,691 @@
+      auto igates = at::matmul(input, params.w_ih.t());
+      auto hgates = at::matmul(hidden, params.w_hh.t());
+      auto result = at::_thnn_fused_gru_cell(igates, hgates, hidden, params.b_ih, params.b_hh);
+      // Slice off the workspace argument (it's needed only for AD).
+      return std::get<0>(result);
+    }
+
+    auto igates = linear(input, params.w_ih, params.b_ih);
+    auto hgates = linear(hidden, params.w_hh, params.b_hh);
+    auto chunked_igates = igates.chunk(3, 1);
+    auto chunked_hgates = hgates.chunk(3, 1);
+
+    auto reset_gate = at::sigmoid(chunked_igates[0] + chunked_hgates[0]);
+    auto input_gate = at::sigmoid(chunked_igates[1] + chunked_hgates[1]);
+    auto new_gate = at::tanh(chunked_igates[2] + reset_gate * chunked_hgates[2]);
+
+    return new_gate + input_gate * (hidden - new_gate);
+  }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// LAYER IMPLEMENTATIONS
+//
+// Layers are scan-like higher-order functions, which take in cells, and
+// transform them to fuctions of signature
+//
+// (io_type input, hidden_type hidden, param_type params) -> (io_type, hidden_type)
+//
+// which can apply the cell over a sequence of inputs, and produce both a new set
+// of hidden states, as well as a concatenated output of each step.
+
+template<typename output_type, typename hidden_type>
+struct LayerOutput {
+  output_type outputs;
+  hidden_type final_hidden;
+};
+
+template<typename io_type, typename hidden_type, typename param_type>
+struct Layer {
+  using output_type = LayerOutput<io_type, hidden_type>;
+  virtual output_type operator()(const io_type& input, const hidden_type& input_hidden, const param_type& params) const = 0;
+};
+
+template<typename hidden_type>
+struct FullLayer : Layer<Tensor, hidden_type, CellParams> {
+  using output_type = typename Layer<Tensor, hidden_type, CellParams>::output_type;
+  using unstacked_output_type = LayerOutput<std::vector<Tensor>, hidden_type>;
+
+  FullLayer(Cell<hidden_type>& cell)
+    : cell_(cell) {};
+
+  unstacked_output_type operator()(std::vector<Tensor> step_inputs, const hidden_type& input_hidden, const CellParams& params) const {
+    std::vector<Tensor> step_outputs;
+    auto hidden = input_hidden;
+    for (size_t i = 0; i < step_inputs.size(); i++) {
+      hidden = cell_(step_inputs[i], hidden, params);
+      step_outputs.push_back(hidden_as_output(hidden));
+    }
+    return {step_outputs, hidden};
+  }
+
+  output_type operator()(const Tensor& inputs, const hidden_type& input_hidden, const CellParams& params) const override {
+    auto unstacked_output = (*this)(inputs.unbind(0), input_hidden, params);
+    return {at::stack(unstacked_output.outputs, 0), unstacked_output.final_hidden};
+  }
+
+  Cell<hidden_type>& cell_;
+};
+
+template<typename dir_hidden_type>
+struct FullBidirectionalLayer : Layer<Tensor, pair_of<dir_hidden_type>, pair_of<CellParams>> {
+  using hidden_type = pair_of<dir_hidden_type>;
```

Comment:
Oh how I wish C++ had dependent types, so you wouldn't have to pass in the `hidden_type` in the template parameters of the parent class :)